File: Server/src/com/mytalk/server/logic/processing/request_processor/world_information/CreateAccount.java
/**
* Filename: CreateAccount.java
* Package: com.mytalk.server.logic.processing.request_processor.world_information
* Author: Nicolo' Toso
* Date: 2013-05-02
*
* Diary:
* Version | Date       | Developer | Changes
* --------+------------+-----------+------------------
* 0.3     | 2013-06-22 |	MF	   | [+] Aggiunta commenti al codice in formato Javadoc
* 0.2	  | 2013-05-13 |    NT	   | [#] Modifica dei nomi di alcune variabili in tutti i metodi al
* 										 fine di renderli più espressivi
* 0.1	  |	2013-05-02 |    NT     | [+] Inserimento classe, oggetti e costruttore     
*
* This software is distributed under GNU/GPL 2.0.
*
* Software licensed to:
* - Zucchetti SRL
* 
* Classe che si occupa di creare un nuovo account del client che ne fa richiesta
*/


package com.mytalk.server.logic.processing.request_processor.world_information;

import com.mytalk.server.data.storage.MD5Converter;
import com.mytalk.server.exceptions.UsernameAlreadyExisting;
import com.mytalk.server.logic.processing.request_processor.GenericRequest;
import com.mytalk.server.logic.shared.ARI;
import com.mytalk.server.logic.shared.WorldPack;
import com.mytalk.server.logic.shared.model_client.PersonalData;

public class CreateAccount extends GenericRequest{
	/**
	 * Costruttore della classe con corpo vuoto poiche' non vi sono campi dati da inizializzare
	 * 
	 * @method +CreateAccount
	 */
	public CreateAccount(){}
	
	/**
	 * Una volta ricevuto il pacchetto ARI, si controlla se è ben formato. Se la risposta è negativa,
	 *  viene restituito un "CorruptedPack". Viceversa, si crea un oggetto di tipo 
	 *  CSDAT1.model.User, contenente le informazioni presenti nel campo dati ari. Inoltre vengono 
	 *  codificati l'indirizzo email e la password, in modo tale da salvarle codificate e non 
	 *  inviarle in chiaro. Infine, chiamando il metodo opportuno del DataAccess (appartenente alla 
	 *  componente CSDAT2), si prosegue con l'aggiunta dell'account, che non implica il login 
	 *  automatico da parte dell'utente. Si ritorna un nuovo pacchetto ARI, specificando che 
	 *  l'aggiunta è andata a buon fine: "SuccessfulCreateAccount". Altrimenti viene catturata la seguente 
	 *  eccezione "UsernameAlreadyExisting" e ritornato un pacchetto con valore richiesta 
	 *  "UsernameAlreadyExistingCreateAccount"
	 *  
	 *  @method +manage
	 *  @param {ARI} ari e' l'oggetto che contiene le informazioni necessarie alla classe per
	 *  poter processare la specifica richiesta e restituire il risultato dell'elaborazione
	 *  @return {ARI}
	 */
	public ARI manage(ARI ari){
		String infoRequest=ari.getInfo();
		ARI response=null;
		WorldPack pack=(WorldPack)conv.convertJsonToJava(infoRequest, WorldPack.class);
		boolean checkPack=checkWorldPackWellFormed(pack);
		if(!checkPack){
			response=new ARI(null,"CorruptedPack",infoRequest);
		}else{
			PersonalData personalData=pack.getWorldPersonalData().getPersonalData();
			String md5=MD5Converter.getHashMD5(personalData.getEmail());
			String pwdHash=MD5Converter.getHashMD5(personalData.getPassword());
			com.mytalk.server.data.model.User user=new com.mytalk.server.data.model.User(personalData.getUsername(), pwdHash,personalData.getEmail(), personalData.getName(), personalData.getSurname(),md5);
			try{
				da.createAccount(user);
				response=new ARI(null, "SuccessfulCreateAccount", infoRequest);
			}catch(UsernameAlreadyExisting uae){
				response=new ARI(null, "UsernameAlreadyExistingCreateAccount", infoRequest);
			}
		}
		return response;
	}

}

File: Client/com/mytalk/client/model/storage/processing/processor/outcoming/user/CreateAccount.js
/**
* Filename: CreateAccount.js
* Package: com.mytalk.client.model.storage.processing.processor.outcoming.user
* Dependencies:  com.mytalk.client.model.storage.processing.processor.outcoming.AbstractOutProcessorProduct
*                com.mytalk.client.model.storage.ARI
* Author: Agostinetto Mattia
* Date: 2013-05-01
*
* Diary:
*
| Version | Date       | Developer | Changes
* --------+------------+-----------+------------------
* 0.1     | 2013-05-01 | MA        | [+] Stesura dell'intera classe
*
* This software is distributed under GNU/GPL 2.0.
*
* Software licensed to:
* - Zucchetti SRL
*
* Processore che viene eseguito quando l'utente si registra
*
*/


MyTalk.processor.CreateAccount = Ember.Object.extend(MyTalk.AbstractOutProcessorProduct,{
  /**
  * Memorizza il nome del processore
  * 
  * @property -name
  * @type {String}
  */
  name: 'CreateAccount',
 /**
  * Il metodo deve creare un nuovo record nel model _DS.PersonalData_ con i dati di registrazione
  *
  * @method +process
  * @param {Object} Stringa JSON che rappresenta il pacchetto ARI
  * @return {Void}
  * @override CCMOD2.processing.processor.outcoming$AbstractOutProcessorProduct$
  */
  process: function (params) {
    var record = MyTalk.PersonalData.find(0).setProperties(params);
    var transaction = record.get('transaction');

    transaction.reopen({
      processor: this
    });

    transaction.commit();
  },
 /**
  * Il metodo deve inviare al server un ARI avente richiesta _CreateAccount_ per poter rendere persistenti i dati contenuti
  * nel pacchetto ARI
  *
  * @method +sendToServer
  * @param {WebSocket} istanza di connessione WebSocket
  * @param {Object} record Call da inviare al server
  * @param {Function} callback che deve essere eseguita quando il pacchetto risulta correttamente inviato al server
  * @return {Void}
  * @override CCMOD2.processing.processor.outcoming$AbstractOutProcessorProduct$
  */
  sendToServer: function (socket, record, onSent) {
    var ARI = new Object();

    ARI.auth = null;
    ARI.req = this.get('name');
    
    var info = new Object();
    info.worldPersonalData = new Object();
    info.worldPersonalData.pd = new Object();
    info.worldPersonalData.pd.username = record.get('username');
    info.worldPersonalData.pd.password = record.get('password');
    info.worldPersonalData.pd.name = record.get('name');
    info.worldPersonalData.pd.surname = record.get('surname');
    info.worldPersonalData.pd.email = record.get('email');
    
    ARI.info = JSON.stringify(info);
    
    socket.send( JSON.stringify(ARI) );
    onSent( this.getProcessorName(), true );
  },
  /**
  * Il metodo deve ritornare l'attributo _name_
  *
  * @method +getProcessorName
  * @return {String}
  * @override CCMOD2.processing.processor.outcoming$AbstractOutProcessorProduct$
  */
  getProcessorName: function () {
    return this.get('name');
  } 
});